synapseclient/client.py

```diff
@@ -1132,6 +1134,75 @@ def _getEntityBundle(self, entity, version=None, bitFlags=0x800 | 0x400 | 0x2 |
 
         return bundle
 
+    def copy(self, entity, parentId, version=None, setProvenance="traceback"):
+        """
+        Copies most recent version of a file to a specified synapse ID.
+
+        :param obj: A synapse ID or entity of a file
+
+        :param parentId: Synapse ID of a folder/project that the file wants to be copied to
+
+        """
+        #Set provenance should take a string (none, traceback, existing)
+        ent = self.get(entity, downloadFile=False, version=version, followLink=False)
+        #CHECK: If file is in the same parent directory (throw an error)
+        search = self.query('select name from file where parentId =="%s"'%parentId)
+        for i in search['results']:
+            if i['file.name'] == ent.name:
+                raise ValueError('Filename exists in directory you would like to copy to, either rename or check if file has already been copied!')
+        profile = self.getUserProfile()
+        # get provenance earlier to prevent errors from being called in the end
+        # If traceback, set activity to old entity
+        if setProvenance == "traceback":
+            act = Activity("Copied file", used=ent)
+        # if existing, check if provenance exists
+        elif setProvenance == "existing":
+            try:
+                act = self.getProvenance(ent.id)
+            except Exception as e:
+                if e.message.find('No activity') >= 0:
+                    act = None
+                else:
+                    raise e
+        elif setProvenance is None:
+            act = None
+        else:
+            raise ValueError('setProvenance must be one of None, existing, or traceback')
+        #Grab file handle createdBy annotation to see the user that created fileHandle
+        #NOTE: May not always be the first index (need to filter to make sure not PreviewFileHandle )
+        fileHandle = self.restGET('/entity/%s/version/%s/filehandles'%(ent.id,ent.versionNumber))
+        # not guaranteed to get preview and regular file in same order, need filter here
+        # External URL's are not 'org.sagebionetworks.repo.model.file.S3FileHandle'
+        if fileHandle['list'][0]['concreteType'] != 'org.sagebionetworks.repo.model.file.PreviewFileHandle':
+            createdBy = fileHandle['list'][0]['createdBy']
+        else:
+            createdBy = fileHandle['list'][1]['createdBy']
```

Comment:
Maybe something like this?

fileHandleList = self.restGET('/entity/%s/version/%s/filehandles'%(ent.id,ent.versionNumber))
for fileHandle in fileHandleList['list']:
    if fileHandle['id'] == ent.dataFileHandleId:
        createdBy = fileHandle['createdBy']
        break
else:
    createdBy = None